### DISCUSSION TITLE
What should be the default type for database ids?

### DISCUSSION THREAD

**Discussion by flybayer** [blitz-js/blitz#554] [General] [answered]
We need to decide what the default id type is for our generated Prisma models (generated via `blitz generate model`.

Traditionally ids have been bigints. This is still the default for Rails today.
 
However, uuids are much more common now with distributed systems because they have many advantages over incremented ids:
 
* Not guessable, so it prevents things like Zoom meeting bombings
* Hides the number of objects in your system, which you may not want people to know
* Much better for distributed systems because the uuid can be generated anywhere without a central authoritative issuer of incremental ids
 
I think we should go with [cuid](https://github.com/ericelliott/cuid) for our default ids. This is already supported by Prisma.
 
The only drawback, if you want to call it that, is that urls are more verbose. But not terribly so.
 
```
// this
/projects/cjld2cjxh0000qzrmn831i7rn/tasks/cjld2cyuq0000t3rmniod1foy
 
// vs this
/projects/9127/tasks/47
 ```
 
What do you all think?

**Comment by merelinguist** [comment]
Personally much prefer `cuid`s too. I think the urls look much better, and obviously protects against security concerns.

We should also have docs around how to do slugged urls.

**Comment by Zeko369** [comment]
## I agree that CUIDs/UUIDs are quite common nowadays and quite useful in bigger systems, BUT

### Honestly I don't think cuid/uuids should be the default since they'll mostly cause confusion and be harder to use in some cases.

#### Regarding your points

> Not guessable, so it prevents things like Zoom meeting bombings

#### -> Use authorization

> Hides the number of objects in your system, which you may not want people to know

#### -> This point is completely valid, but for some stuff like that you can just use slugs, or if you'd want UUID/CUID

> Much better for distributed systems because the uuid can be generated anywhere without a central authoritative issuer of incremental ids

#### -> Since we're still storing all of this in a single DB, this isn't really and issue

#### e.g. in console

`> Product.find("cjld2cjxh0000qzrmn831i7rn")` vs `> Product.find(2)`

Also harder to find the next product from this

#### URLS
```
Seems quite long and hard to read
/projects/cjld2cjxh0000qzrmn831i7rn/tasks/cjld2cyuq0000t3rmniod1foy

Rails way
/projects/9127/tasks/47

Slugs for some stuff
/projects/foobar/tasks/47
```

**Comment by MrLeebo** [comment]
There are also drawbacks when it comes to database performance. UUIDs are slower to compare and take up more memory, so you can fit fewer of them on a page.

**Reply by flybayer** [reply]
True, but I think realistically for the vast majority of cases the difference is effectively zero.

If someone is building an app large enough that this matters, they will know to change the id to bigint anyways.

**Comment by lachlanjc** [comment]
URLs, & therefore IDs, are a part of an app’s UX—if GitHub repositories were at /repositories/18474385 or if Slack channels were at /channels/general, they’d feel very different to use. I don’t have strong feelings one or another, but I’d love for this to be easy to customize per-app.

**Reply by flybayer** [reply]
For cases where the url matters, you need to use slugs anyways. So this doesn't really affect whether ids are strings or ints. Github does this (`github.com/blitz-js/blitz`) and for all we know, the repo ids could be either strings or integers.

**Comment by dwightwatson** [comment]
I'll put a vote in for integers. UUIDs being more common with distributed systems makes sense, but if under the hood we're using a regular database like Postgres then is it actually solving a problem?

IDs are nicer, easier to reason with, and result in shorter URLs. I agree with @Zeko369 that doing authorization is a better approach than hoping a UUID won't be guessed. And I think needing to hide the number of records your app has is more of an outlier than something that should be addressed out of the box.

What's the default generated by Prisma (I presume it's an integer ID) and are any of the arguments above really enough justification to change that default?

**Reply by flybayer** [reply]
Prisma doesn't have a default

**Comment by johhansantana** [comment]
I started a discussion not so long ago about the differences between using either in the Prisma repo

https://github.com/prisma/prisma/discussions/2515

By the answer there I would say bigint should be the default and probably give the user the ability to change this in the config maybe?

**Comment by sijad** [comment]
I also think bigint should be the default

> * Not guessable, so it prevents things like Zoom meeting bombings
> * Hides the number of objects in your system, which you may not want people to know

both can be solved using something like https://hashids.org/

**Reply by flybayer** [reply]
Wow, hashids looks pretty cool!

**Comment by jportela** [comment]
I think `BigInt` is simpler to reason about than a `cuid`, so I’d recommend it for Blitz as a default. We could later have a guide to allow users to change the type for`cuid` or `uuid` where we explain why we would use them over BigInt

**Reply by johnletey** [reply]
@jportela I 💯agree with this!

**Comment by cr101** [comment]
I prefer incremental IDs as I find it easier to work with, query the database and test the app.

Does Blitz allow for using slugs for URLs? Using slugs instead of random strings/ints in URLs is way better for SEO.

**Reply by flybayer** [reply]
Yeah you can definitely use slugs!

**Comment by flybayer** [comment, marked as answer]
Ok, y'all have spoken! We'll stick with default integer IDs :)
